fem/tbilininteg.hpp

```diff
@@ -554,6 +554,1126 @@ struct TDiffusionKernel<3,3,complex_t>
+   static inline MFEM_ALWAYS_INLINE
+   void MultAssembled(const int k, const TMatrix<qpts,6,complex_t> &A,
+                      S_data_t &R)
+   {
+      const int M = S_data_t::eval_type::qpts;
+      const int NC = S_data_t::eval_type::vdim;
+      MFEM_STATIC_ASSERT(qpts == M, "incompatible dimensions");
+      MFEM_FLOPS_ADD(15*M*NC);
+      for (int i = 0; i < M; i++)
+      {
+         const complex_t A11 = A(i,0);
+         const complex_t A21 = A(i,1);
+         const complex_t A31 = A(i,2);
+         const complex_t A22 = A(i,3);
+         const complex_t A32 = A(i,4);
+         const complex_t A33 = A(i,5);
+         for (int j = 0; j < NC; j++)
+         {
+            const complex_t x1 = R.curl_qpts(i,0,j,k);
+            const complex_t x2 = R.curl_qpts(i,1,j,k);
+            const complex_t x3 = R.curl_qpts(i,2,j,k);
+            R.curl_qpts(i,0,j,k) = A11*x1 + A21*x2 + A31*x3;
+            R.curl_qpts(i,1,j,k) = A21*x1 + A22*x2 + A32*x3;
+            R.curl_qpts(i,2,j,k) = A31*x1 + A32*x2 + A33*x3;
+         }
+      }
+   }
+};
+
+// mass kernel for RT Finite Element
+
+// complex_t - type for the assembled data
+template <int SDim, int Dim, typename complex_t>
+struct THDivMassKernel;
+
+// mass kernel for RT Finite Element kernel in 2D
+template <typename complex_t>
+struct THDivMassKernel<2,2,complex_t>
+{
+   typedef complex_t complex_type;
+
+   // needed for the TElementTransformation::Result class
+   static const bool uses_Jacobians = true;
+
+   // needed for the FieldEvaluator::Data class
+   static const bool in_values  = true;
+   static const bool in_divs    = false;
+   static const bool out_values = true;
+   static const bool out_divs   = false;
+
+   // Partially assembled data type for one element with the given number of
+   // quadrature points. This type is used in partial assembly, and partially
+   // assembled action. Stores one symmetric 2 x 2 matrix per point.
+   template <int qpts>
+   struct p_asm_data { typedef TMatrix<qpts,3,complex_t> type; };
+
+   // Partially assembled data type for one element with the given number of
+   // quadrature points. This type is used in full element matrix assembly.
+   // Stores one general (non-symmetric) 1 x 1 matrix per point.
+   template <int qpts>
+   struct f_asm_data { typedef TTensor3<qpts,2,2,complex_t> type; };
+
+   template <typename IR, typename coeff_t, int NE>
+   struct CoefficientEval
+   {
+      typedef typename IntRuleCoefficient<IR,coeff_t,NE>::Type Type;
+   };
+
+   // Method used for un-assembled (matrix free) action.
+   // Jt        [M x Dim x SDim x NE] - Jacobian transposed, data member in F
+   // Q                               - CoefficientEval<>::Type
+   // q                               - CoefficientEval<>::Type::result_t
+   // val_qpts [M x SDim x NC x NE]  - in/out data member in R
+   //
+   // curl_qpts = (w/det(J))J val_qpts
+   template <typename T_result_t, typename Q_t, typename q_t,
```

Comment:
Same remark as for Hcurl, is there a reason why the coefficients are curl_qpts = (w/det(J))J val_qpts, and not simply w*det(J)?

Reply:
Ok, i see that the continuity mapping appears here. Shouldn't we have a more general MassIntegrator then?